### PR TITLE
Facility selection: discard stale location updates

### PR DESCRIPTION
**Pull request by saket**
This PR brings brings a portion of changes towards this feature:
https://www.pivotaltracker.com/story/show/158963954

Major changes:
 - Discard stale location updates that are older than 10m
 - Hide suggested facilities once the user starts searching facilities
 - Read user location only when permission is granted

### CODE REVIEW

File: app/src/main/java/org/simple/clinic/facility/change/FacilityListItemBuilder.kt [hunk 1]
```diff
@@ -27,7 +27,7 @@ object FacilityListItemBuilder {
     }
 
     val listItems = mutableListOf<FacilityListItem>()
-    if (nearbyFacilityItems.isNotEmpty()) {
+    if (searchQuery.isBlank() && nearbyFacilityItems.isNotEmpty()) {
```
Comment (rakshakhegde) [inline app/src/main/java/org/simple/clinic/facility/change/FacilityListItemBuilder.kt]: Is there a corresponding test for this change?

File: app/src/test/java/org/simple/clinic/registration/facility/RegistrationFacilitySelectionScreenControllerTest.kt [hunk 2]
```diff
@@ -139,21 +143,27 @@ class RegistrationFacilitySelectionScreenControllerTest {
 
   @Test
   fun `when screen is started then location should only be read once`() {
```
Comment (rakshakhegde) [inline app/src/test/java/org/simple/clinic/registration/facility/RegistrationFacilitySelectionScreenControllerTest.kt]: The test for checking if stale location has been discarded seems to be missing. Can you please add this?

File: app/src/main/java/org/simple/clinic/registration/facility/FacilitiesAdapter.kt [hunk 3]
```diff
@@ -137,5 +138,7 @@ class FacilityOptionViewHolder(rootView: View) : ViewHolder(rootView) {
             address.state)
       }
     }
+
+    dividerView.visibility = if (facilityOption.showBottomDivider) View.VISIBLE else View.GONE
```
[the remark this answers is not in the document]
Comment (saket) [reply app/src/main/java/org/simple/clinic/registration/facility/FacilitiesAdapter.kt]: Done

[the remark this answers is not in the document]
Comment (saket) [reply app/src/main/java/org/simple/clinic/registration/facility/FacilitiesAdapter.kt]: Side note: that sounds like a getter and not a setter

File: app/src/main/java/org/simple/clinic/facility/change/FacilityListItemBuilder.kt [hunk 4]
```diff
@@ -77,6 +82,10 @@ object FacilityListItemBuilder {
       FacilityOption.Address.WithStreet(street = facility.streetAddress, district = facility.district, state = facility.state)
     }
 
-    return FacilityOption(facility = facility, name = highlightedName, address = fullAddress)
+    return FacilityOption(
+        facility = facility,
+        name = highlightedName,
+        address = fullAddress,
+        showBottomDivider = isLastItemInSection.not())
```
[the remark this answers is not in the document]
Comment (saket) [reply app/src/main/java/org/simple/clinic/facility/change/FacilityListItemBuilder.kt]: Sure

File: app/src/main/java/org/simple/clinic/facility/change/FacilityListItemBuilder.kt (hunk 1 above)
Comment (saket) [reply app/src/main/java/org/simple/clinic/facility/change/FacilityListItemBuilder.kt]: There is, but I accidentally checked it in the previous commit. I'll fix this.

File: app/src/test/java/org/simple/clinic/registration/facility/RegistrationFacilitySelectionScreenControllerTest.kt (hunk 2 above)
Comment (saket) [reply app/src/test/java/org/simple/clinic/registration/facility/RegistrationFacilitySelectionScreenControllerTest.kt]: Good catch! I've added a test.

Comment (saket) [reply app/src/test/java/org/simple/clinic/registration/facility/RegistrationFacilitySelectionScreenControllerTest.kt]: What do you think?

https://github.com/simpledotorg/simple-android/blob/6055b777dc2ec9d372c2c2a010e3812c1ec66e5f/app/src/test/java/org/simple/clinic/registration/facility/RegistrationFacilitySelectionScreenControllerTest.kt#L172

**Comment by saket** [comment]
@rakshakhegde feedback addressed!

**Comment by rakshakhegde** [comment]
@saket 2 tests seem to have failed

File: app/src/main/java/org/simple/clinic/facility/change/FacilityListItemBuilder.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (saket) [reply app/src/main/java/org/simple/clinic/facility/change/FacilityListItemBuilder.kt]: On a second thought, I don't think we should do this. I think there's a difference between `default` and `optional` value. I think we should always keep values explicit in favor of discoverability unless they're optional.

**Comment by saket** [comment]
@rakshakhegde whoops. I've fixed the test.

File: app/src/main/java/org/simple/clinic/facility/change/FacilityListItemBuilder.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (saket) [reply app/src/main/java/org/simple/clinic/facility/change/FacilityListItemBuilder.kt]: On a third thought, let's discuss this with the team :p

**Pull request merged**